numba_cuda/numba/cuda/cudadrv/driver.py

```diff
@@ -2546,38 +2586,58 @@ def launch_kernel(cufunc_handle,
                               extra)
 
 
-if USE_NV_BINDING:
-    jitty = binding.CUjitInputType
-    FILE_EXTENSION_MAP = {
-        'o': jitty.CU_JIT_INPUT_OBJECT,
-        'ptx': jitty.CU_JIT_INPUT_PTX,
-        'a': jitty.CU_JIT_INPUT_LIBRARY,
-        'lib': jitty.CU_JIT_INPUT_LIBRARY,
-        'cubin': jitty.CU_JIT_INPUT_CUBIN,
-        'fatbin': jitty.CU_JIT_INPUT_FATBINARY,
-    }
-else:
-    FILE_EXTENSION_MAP = {
-        'o': enums.CU_JIT_INPUT_OBJECT,
-        'ptx': enums.CU_JIT_INPUT_PTX,
-        'a': enums.CU_JIT_INPUT_LIBRARY,
-        'lib': enums.CU_JIT_INPUT_LIBRARY,
-        'cubin': enums.CU_JIT_INPUT_CUBIN,
-        'fatbin': enums.CU_JIT_INPUT_FATBINARY,
-    }
-
-
 class Linker(metaclass=ABCMeta):
     """Abstract base class for linkers"""
 
     @classmethod
-    def new(cls, max_registers=0, lineinfo=False, cc=None):
-        if config.CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY:
-            return MVCLinker(max_registers, lineinfo, cc)
-        elif USE_NV_BINDING:
-            return CudaPythonLinker(max_registers, lineinfo, cc)
+    def new(cls,
+            max_registers=0,
+            lineinfo=False,
+            cc=None,
+            lto=None,
+            additional_flags=None
+            ):
+
+        driver_ver = driver.get_version()
+        if (
+            config.CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY
+            and driver_ver >= (12, 0)
+        ):
+            raise ValueError(
+                "Use ENABLE_PYNVJITLINK for CUDA >= 12.0 MVC"
+            )
+        if config.ENABLE_PYNVJITLINK and driver_ver < (12, 0):
+            raise ValueError(
+                "Use CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY "
+                "for CUDA < 12.0 MVC"
+            )
+        if (
+            config.CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY
+            and config.ENABLE_PYNVJITLINK
+        ):
+            raise ValueError(
+                "can't set both config.ENABLE_PYNVJITLINK "
+                "and config.CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY "
+                "at the same time"
+            )
```

Comment:
I think this check could be pushed up to the top of the file around line 86 where the check for pynvjitlink being enabled / imported is - this would allow us to report a setup issue to the user quickly instead of waiting for them to compile / link something.